awscrt/http.py

```diff
@@ -126,24 +146,34 @@ def new(cls,
 
             if not bootstrap:
                 bootstrap = ClientBootstrap.get_or_create_static_default()
-
-            connection = cls()
-            connection._host_name = host_name
-            connection._port = port
+            shutdown_future = None
 
             def on_connection_setup(binding, error_code, http_version):
-                if error_code == 0:
-                    connection._binding = binding
-                    connection._version = HttpVersion(http_version)
-                    future.set_result(connection)
-                else:
+                if expected_version and expected_version != http_version:
+                    # unexpected protocol version
+                    # AWS_ERROR_HTTP_UNSUPPORTED_PROTOCOL
+                    future.set_exception(awscrt.exceptions.from_code(2060))
+                    return
+                if error_code != 0:
                     future.set_exception(awscrt.exceptions.from_code(error_code))
+                    return
+                if http_version == HttpVersion.Http2:
+                    connection = Http2ClientConnection()
+                else:
+                    connection = HttpClientConnection()
+                connection._host_name = host_name
+                connection._port = port
 
-            # on_shutdown MUST NOT reference the connection itself, just the shutdown_future within it.
-            # Otherwise we create a circular reference that prevents the connection from getting GC'd.
-            shutdown_future = connection.shutdown_future
+                connection._binding = binding
+                connection._version = HttpVersion(http_version)
+                nonlocal shutdown_future
+                shutdown_future = connection.shutdown_future
```

Comment:
debatable: maybe move to the [_Core pattern](https://github.com/awslabs/aws-crt-python/blob/af06aa55085c437351b32a5870f48fd6a1320f97/guides/dev/writing-code.md#option-2---private-core-class)? this is getting pretty fancy